drivers/t5838/t5838_aad_mode.c

```diff
@@ -0,0 +1,486 @@
+{
+	if (g_drv_data->int_handled) {
+		return;
+	}
+	g_drv_data->int_handled = true;
+	int err = 0;
+	err = gpio_pin_interrupt_configure_dt(&g_drv_cfg->wake, GPIO_INT_DISABLE);
+	if (err) {
+		LOG_ERR("gpio_pin_interrupt_configure, err: %d", err);
+		return;
+	}
+	LOG_INF("WAKE PIN INTERRUPT");
+
+	if (g_drv_data->wake_handler) {
+		g_drv_data->wake_handler(dev);
+	}
+}
+
+int t5838_reset(const struct device *dev)
+{
+
+	const struct t5838_aad_drv_cfg *drv_cfg = dev->config;
+	struct t5838_aad_drv_data *drv_data = dev->data;
+	int err;
+	if (drv_cfg->micen_available) {
+		err = gpio_pin_set_dt(&drv_cfg->micen, 0);
+		if (err < 0) {
+			LOG_ERR("gpio_pin_set_dt, err: %d", err);
+			return err;
+		}
+
+		drv_data->aad_unlocked = false;
+		drv_data->aad_enabled_mode = T5838_AAD_SELECT_NONE;
+
+		/* Wait for device to power down and reapply power to reset it */
+		k_sleep(K_MSEC(T5838_RESET_TIME_MS));
+		err = gpio_pin_set_dt(&drv_cfg->micen, 1);
+		if (err < 0) {
+			LOG_ERR("gpio_pin_set_dt, err: %d", err);
+			return err;
+		}
+		return 0;
+	}
+	return -ENOTSUP;
+}
+
+int t5838_aad_wake_clear(const struct device *dev)
+{
+	const struct t5838_aad_drv_cfg *drv_cfg = dev->config;
+	struct t5838_aad_drv_data *drv_data = dev->data;
+	int err;
+	if (drv_data->cb_configured) {
+		err = gpio_pin_interrupt_configure_dt(&drv_cfg->wake, GPIO_INT_EDGE_TO_ACTIVE);
+		if (err) {
+			LOG_ERR("gpio_pin_interrupt_configure, err: %d", err);
+			return err;
+		}
+	}
+	drv_data->int_handled = false;
+	return 0;
+}
+
+void t5838_aad_wake_handler_set(const struct device *dev, t5838_wake_handler_t handler)
+{
+	struct t5838_aad_drv_data *drv_data = dev->data;
+	drv_data->wake_handler = handler;
+}
+
+/**
+ * @brief Function for setting AAD mode from provided array of address data pairs.
+ *
+ * @param[in] dev Pointer to the device structure for the driver instance.
+ * @param[in] write_array Array of address data pairs with setup sequence to write to device.
+ * @param[in] write_array_size Size of write_array.
+ */
+int prv_aad_mode_set(const struct device *dev, struct t5838_address_data_pair *write_array,
+		     uint8_t write_array_size)
+{
+	const struct t5838_aad_drv_cfg *drv_cfg = dev->config;
+	struct t5838_aad_drv_data *drv_data = dev->data;
+	int err;
+
+	/** Make sure there is no PDM transfer in progress, and we can take clock signal */
+	struct t5838_drv_data *pdm_data = drv_cfg->pdm_dev->data;
+	if (pdm_data->active) {
+		LOG_ERR("Cannot write to device while pdm is active");
+		return -EBUSY;
+	}
+	if (drv_data->aad_unlocked == false) {
+		err = prv_aad_unlock_sequence(dev);
+		if (err < 0) {
+			LOG_ERR("error writing aad unlock sequence, err: %d", err);
+			return err;
+		}
+	}
+
+	err = prv_multi_reg_write(dev, write_array, write_array_size);
+	if (err < 0) {
+		LOG_ERR("prv_t5838_multi_reg_write, err: %d", err);
+		return err;
+	}
+
+	err = prv_enter_sleep_with_AAD(dev);
+	if (err < 0) {
+		LOG_ERR("prv_t5838_reg_write, err: %d", err);
+		return err;
+	}
```

Comment:
that is correct, but I think it should be done the other way around and check and make sure clock was not taken by PDM peripheral?

Reply:
On the second thought, you are right. Since we are calling this immediately after disabling it, checking for active condition will fail since PDM event handler gets set later (which might not be best idea though).